Updated PSoC 6 implementation to use hal_infineon/mtb-pdl-cat1

npal-cy opened the pull request:
Updated old PSoC 6 implementation to use hal_infineon/mtb-pdl-cat1

This PR is reflected changes according to Task#1 for RFC#42883

This PR depend on: https://github.com/zephyrproject-rtos/hal_infineon/pull/2
NOTE: after merge hal_infineon, the west.yaml should be finally updated.

npal-cy committed:
soc: arm: cypress: Update PSoC 6 implementation to use mtb-pdl-cat1

Updated PSoC 6 implementation to use hal_infineon/mtb-pdl-cat1

This PR is reflected changes according to Task#1 for RFC#42883.

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

nandojve commented on west.yml:
```diff
@@ -61,11 +61,6 @@ manifest:
       path: modules/hal/atmel
       groups:
         - hal
-    - name: hal_cypress
```
Hi @npal-cy ,

You need add hal_infineon at same time you drop hal_cypress.
Remember that you need add your PR number at revision.

revision: pull/2/head

npal-cy replied:
correct. in description i added note about dependency on: https://github.com/zephyrproject-rtos/hal_infineon/pull/2.
So when we merge hal_infineon/pull/2, i will add one more commit with update hal_infineon in west.yaml

nandojve replied:
May I suggest you create another PR only with this dependency?
Maybe we can get that faster if that is the only necessary dependency right now.

carlescufi replied:
This is the PR that should get https://github.com/zephyrproject-rtos/hal_infineon/pull/2 into the tree @nandojve. Later we will then continue with https://github.com/zephyrproject-rtos/zephyr/pull/44211, but first let's get this merged.

nandojve reviewed (changes requested):
Hi @npal-cy ,

Thank you for this change!
It seems that PDL API not changed too much since version 3.1.0. Is that correct?

npal-cy commented:
> It seems that PDL API not changed too much since version 3.1.0. Is that correct?

yes, the interface was not changes, but 3.1.0 is a little bit old (from PSoC Creator). 

gpio  1.20.1 ->  1.50.1  [gpio changelog](https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__gpio.html) 

uart/spi   2.20 -> 2.80   [scb changelog](https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual/html/group__group__scb.html) 

you can find change log for each driver in online documentation https://infineon.github.io/mtb-pdl-cat1/pdl_api_reference_manual

nandojve reviewed (changes requested):
Please, remove `old` word reference on the commit message.

nandojve reviewed (changes requested):
.

nandojve commented on west.yml:
```diff
@@ -76,7 +76,7 @@ manifest:
         - hal
     - name: hal_infineon
-      revision: f1fa8241f8786198ba41155413243de36ed878a5
+      revision: pull/2/head
```
This should be squashed with previous commit to allow bisect.

npal-cy committed:
modules: hal_infineon: added initial Kconfig and CMakeLists.txt

- added initial Kconfig
- added initial CMakeLists.txt (root, core-lib,  mtb-pal-cat1)
- updated module, driver and soc to use CONFIG_USE_INFINEON_xx
  defines from modules/hal_infineon/Kconfig

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
west: Update hal_infineon and hal_cypress

- Update revision hal_infineon to 4af06965f57ba1e7d170e6a97d24c33785543a8c
- Remove hal_cypress

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

gmarull commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+
+if(CONFIG_ADC)
+  zephyr_library_sources(${pdl_drv_dir}/source/cy_sysanalog.c)
+endif()
```
please don't use generic symbols to enable conditional compilation, see how other HALs do this using hidden selectable symbols, STM32 or GD32 HALs. Seems like this is addressed in next commit: just squash changes, there's no need to keep a wrong intermediate solution in git history.

npal-cy replied:
yes i done in https://github.com/zephyrproject-rtos/zephyr/pull/43494/commits/c75442ac81197b8a3936e09d0b73836eaab1c6ba. I will squash.

npal-cy replied:
squashed.

gmarull commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+zephyr_library_sources(${pdl_drv_dir}/source/cy_syspm.c)
+zephyr_library_sources(${pdl_drv_dir}/source/cy_systick.c)
+zephyr_library_sources(${pdl_drv_dir}/source/cy_trigmux.c)
+zephyr_library_sources(${pdl_drv_dir}/source/cy_wdt.c)
```
some of these sources seems to be optional, e.g. `cy_gpio.c` should only be compiled if the Cypress GPIO driver selects to use HAL module.

npal-cy replied:
cy_gpio.c is used in almost all hal modules in cyhal_pin structure. so i added it to common part (without conditions).

we can add this under USE_INFINEON_GPIO, and then enable mandatory it in SOC_FAMILY_PSOC6
but looks likes if we go with that way, we need to add USE_INFINEON_XXX for all common part.

I would like to file internal ticket and update in scope of another PR (e.g. zephyr/pull/44211). Please provide your thoughts.

gmarull commented on soc/arm/cypress/Kconfig:
```diff
@@ -29,6 +29,7 @@ config SOC_PSOC6_M0_ENABLES_M4
 
 config SOC_FAMILY_PSOC6
 	bool
+	select USE_INFINEON_WDT
```
this likely needs to be selected by a wdt driver. note that HAL should not be "smart" by enabling devices on its own, unless there's a reason.

npal-cy replied:
it required by SystemInit (https://github.com/zephyrproject-rtos/hal_infineon/blob/4af06965f57ba1e7d170e6a97d24c33785543a8c/mtb-pdl-cat1/devices/COMPONENT_CAT1A/templates/COMPONENT_MTB/COMPONENT_CM4/system_psoc6_cm4.c#L157), where we Unlock and disable WDT during PSoC 6 init.

npal-cy replied:
updated

npal-cy commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+zephyr_include_directories(${pdl_dev_cat1a_dir}/templates/COMPONENT_MTB)
+
+zephyr_library_sources(${pdl_dev_cat1a_dir}/source/cy_device.c)
+zephyr_library_sources(${pdl_drv_dir}/source/TOOLCHAIN_GCC_ARM/cy_syslib_gcc.S)
```
those functions are used by PDL functions.

similar in another vendors: 
__get_PRIMASK/__set_PRIMASK - in stm32
DisableGlobalIRQ/EnableGlobalIRQ - in nxp

carlescufi replied:
Yes, this is literally `irq_lock()` and `irq_unlock()` in Zephyr

gmarull replied:
I see many drivers making use of `Cy_SysLib_EnterCriticalSection()` and `Cy_SysLib_ExitCriticalSection()`, could you please explain why? could that have an impact when using ZLI?

carlescufi replied:
@npal-cy @gmarull Right now I’d go with the Infineon critical sections until you can figure out how to abstract them, like nrfx does:
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrfx/nrfx_glue.h

npal-cy replied:
The drivers use critical section during operation with hw, where the sequences is critical and can not be interrupted by side interrupt event. 
Regarding to impact, it really depends on use case.  The driver's run-time functions can uses critical section to cover some register access (short time operation), some complex sequences under critical section are mostly in init functions, which should not impact ZLI.

gmarull replied:
To be honest, I'm not happy on introducing HALs that do things _their own way_. There should probably be some rules on what HALs can/cannot do. Otherwise, things like https://github.com/zephyrproject-rtos/zephyr/pull/42395 happen, ie working around HALs in Zephyr. Similar things happen for clock control: many SoCs call a "init" function from the HAL that does all the magic, bypassing any kind of good practices.

To summarize, low level HALs _can_ be useful, higher level ones tend to be inefficient and cause troubles because they do _too much_. This is likely because they were designed in a bare-metal/FreeRTOS mindset, where you basically have no infrastructure.

Modules are nowadays an entry point for code that doesn't get the same level of review that upstream code does, but still, often it is code that is fundamental to build an image.

npal-cy replied:
@gmarull, i do not agree about you concerns for vendor implementation of critical section. We are talking about PDL - low level layer, which directly works with HW, and knows when need to use critical section for particular specific HW and how implement in to achieve all needs.. 

Zephyr can not require to use Zephyr implementation (irq_lock/irq_unlock) for low level (vendor specific) layer, because it does not know all HW specifics.. Zephyr can require to do this for Zephyr drivers, app, or other modules, which are upper vendor peripheral drivers (low level).

I can not guarantee that our HW will work properly with irq_lock,  because irq_lock allows to interrupt the locked by irq_lock  code if ZLI enabled (it uses BASEPRI instead PRIMASK).

carlescufi replied:
So what will happen with the current implementation of the `Cy_SysLib_*CriticalSection` functions, which just use the Armv6-M `cpsid/e` basic `PRIMASK` access is that:

- Infineon HAL drivers will disable/enable ZLIs. This is unfortunate, but you could just add a Kconfig that disallows `CONFIG_ZERO_LATENCY_IRQS` with this SoC
- Infineon HAL drivers will disable the highest prio (prio 0) which is used by Zephyr's fault handling code.  This means that if a fault is triggered inside one of those critical sections then it will not be handled

All in all I don't see this as a show stopper to get this merged, but it certainly needs to be fixed at some point in the near future.

npal-cy commented on modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+
+if(CONFIG_USE_INFINEON_TRNG)
+  zephyr_library_sources(${pdl_drv_dir}/source/cy_crypto.c)
+  zephyr_library_sources(${pdl_drv_dir}/source/cy_crypto_core_trng_v1.c)
+  zephyr_library_sources(${pdl_drv_dir}/source/cy_crypto_core_trng_v2.c)
```
no. it supports different version of Crypto HW.
if you look at PDL code, the final user API e.g. Cy_Crypto_Core_Trng looks like:
```
__STATIC_INLINE cy_en_crypto_status_t Cy_Crypto_Core_Trng(CRYPTO_Type *base,
                                             uint32_t  GAROPol,
                                             uint32_t  FIROPol,
                                             uint32_t  max,
                                             uint32_t *randomNum)
{
    cy_en_crypto_status_t tmpResult;
    if (CY_CRYPTO_V1)
    {
        tmpResult = Cy_Crypto_Core_V1_Trng(base, GAROPol, FIROPol, max, randomNum);
    }
    else
    {
        tmpResult = Cy_Crypto_Core_V2_Trng(base, GAROPol, FIROPol, max, randomNum);
    }
    return (tmpResult);
}
```

npal-cy commented on modules/hal_infineon/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+## Add core-lib sources for CAT1 devices
+if (CONFIG_SOC_FAMILY_PSOC6)
+  add_subdirectory(core-lib)
+endif()
```
updated.

nandojve reviewed (approved):
LGTM!

Thank you Nazar.

carlescufi reviewed (approved):
The discussion over critical section implementation is now resolved.

stephanosio reviewed (commented):
This is causing the following warning in CMake:

```
$ west build -b native_posix samples/synchronization/ --pristine
...
CMake Warning at /home/stephanos/Dev/zephyrproject/zephyr/CMakeLists.txt:820 (message):
  No SOURCES given to Zephyr library: modules_hal_infineon

  Excluding target from build.
```

stephanosio commented on modules/hal_infineon/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2022 Cypress Semiconductor Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_named(modules_hal_infineon)
```
This is causing the CMake warning.

`modules_hal_infineon` library is defined even if none of below subdirectories is added; so, when not building for Infineon/Cypress devices, the source list for this library will be empty.

This needs to go into `if(CONFIG_HAS_XMCLIB OR CONFIG_SOC_FAMILY_PSOC6)`.

Another option is to add `zephyr_library_property(ALLOW_EMPTY TRUE)`.

npal-cy replied:
Hi @stephanosio , thanks for catching this. I am working on PR with fix.

npal-cy replied:
PR: https://github.com/zephyrproject-rtos/zephyr/pull/45671
